### PR TITLE
Mega update

### PR DESCRIPTION
**Pull request by arnar-peturs**
Run php stan to max level

### CODE REVIEW

**Comment by sunkan** [comment]
@arnar-peturs 

Can you add the files needed for github actions?

You can copy the files from here https://github.com/stefna/islandis-auth/tree/master/.github/workflows

Just remove php7.1 - 7.3 from the matrixes

Also please run `vendor/bin/phpcs -n --standard=vendor/stefna/codestyle/library.xml src/`

File: composer.json [hunk 1]
```diff
@@ -18,8 +18,12 @@
...
     "php-http/message": "^1.0",
     "guzzlehttp/psr7": "^1.0",
-    "phpunit/phpunit": "5.*",
-    "monolog/monolog": "^1.0"
+    "monolog/monolog": "^1.0",
+    "roave/security-advisories": "dev-latest",
+    "phpunit/phpunit": "^9.0",
+    "phpstan/phpstan": "^1.0",
+    "stefna/codestyle": "^1.0",
+    "stefna/release-tools": "^1.5"
```
Comment (sunkan) [inline composer.json]: Remove `stefna/release-tools` it's not a public package

**Comment by arnar-peturs** [comment]
@sunkan 
Did add the workflow things. You might want to take a look the failing checks

File: tests/Api/Campaigns/ContentsTest.php [hunk 2]
```diff
@@ -14,11 +15,14 @@ class ContentsTest extends AbstractTestCase
 
 	public function testGetWorks(): void
 	{
+		/** @var Content $ret */
 		$ret = $this->getClient()->campaigns()->contents(self::CAMPAIGN_ID)->get();
 		$this->assertInstanceOf(Content::class, $ret);
 		$this->assertStringContainsString('<title>TestSubject 00</title>', $ret->archiveHtml);
 		$this->assertStringContainsString('<title>*|MC:SUBJECT|*</title>', $ret->html);
-		$this->assertNull($ret->plainText);
+		$rp = new ReflectionProperty(Content::class, 'plainText');
```
Comment (sunkan) [inline tests/Api/Campaigns/ContentsTest.php]: Short answer too this is NO

I think the correct solution to this is to change ` Stefna\Mailchimp\Model\Campaign\Content`

from
```php
public string $plainText;
```
to
```php
public ?string $plainText = null;
```

And I think this is the solution for all places you get `must not be accessed before initialization` errors

**Review by sunkan — CHANGES_REQUESTED** [review]

File: src/Client.php [hunk 3]
```diff
@@ -1,90 +1,82 @@
...
+	public function __construct(
+		HttpClient $httpClient,
+		string $apiKey,
+		?string $apiEndpoint = null,
+		?MessageFactory $messageFactory = null
+	) {
 		$this->httpClient = $httpClient;
 		$this->apiKey = $apiKey;
 		$this->apiEndpoint = $apiEndpoint ?: $this->createApiEndpoint($apiKey);
 		$this->messageFactory = $messageFactory ?: MessageFactoryDiscovery::find();
```
Comment (bix0r) [inline src/Client.php:39]: @sunkan Shouldn't we try to switch this out for something that is not deprecated?

**Comment by bix0r** [comment]
I think this is starting to look good :) 

Perhaps we should force the code-style in ci also?

**Comment by arnar-peturs** [comment]
@sunkan @bix0r 
Ready for final review?

File: src/Client.php (hunk 3 above)
Comment (sunkan) [reply src/Client.php:39]: I think we should merge this then I can make a separate pr with that change

File: src/Client.php [hunk 4]
```diff
@@ -214,55 +232,64 @@ public function response(ResponseInterface $response)
 	}
 
 	/**
-	 * @return array
+	 * @return string[]
```
Comment (sunkan) [inline src/Client.php]: @arnar-peturs this type hint is wrong it should be `array<string, string>`

`string[]` means `array<int, string>`

File: src/Model/Campaign/ABTestOptions.php [hunk 5]
```diff
@@ -8,78 +9,58 @@ class ABTestOptions extends AbstractData
...
+	//public $combinations;
 
 	/**
 	 * Descriptions of possible email contents.
-	 *
-	 * @var array
+	 * NOT IN https://mailchimp.com/developer/marketing/api/campaigns/update-campaign-settings/
 	 */
-	public $contents;
+	//public $contents;
```
Comment (sunkan) [inline src/Model/Campaign/ABTestOptions.php]: If it's not in the api shouldn't it be removed?

File: src/Model/Campaign/CampaignList/SegmentOptions.php [hunk 6]
```diff
@@ -8,24 +9,19 @@ class SegmentOptions extends AbstractData
 {
 	/**
 	 * An array of segment conditions.
+	 * https://mailchimp.com/developer/marketing/docs/alternative-schemas/#segment-condition-schemas
 	 *
-	 * @var array
+	 * @var array<int|array<string, string>>
```
Comment (sunkan) [inline src/Model/Campaign/CampaignList/SegmentOptions.php]: Did you mean `array<int, array<string, string>>`?

**Review by sunkan — CHANGES_REQUESTED** [review]

**Review by sunkan — APPROVED** [review]

**Review by bix0r — APPROVED** [review]
LGTM

**Pull request merged**